### PR TITLE
sort feed by last_activity_at, behind experiment

### PR DESCRIPTION
**Pull request by camhashemi**
@leogrim @ryanpbrewster @andrewconner

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 1]
```diff
@@ -675,6 +676,64 @@ class KeepRepoImpl @Inject() (
...
+    val ktl_JOIN_lm_WHERE_THIS_USER = s"""keep_to_library ktl inner join library_membership lm on (ktl.library_id = lm.library_id) where lm.state = 'active' and ktl.state = 'active' and lm.user_id = $userId"""
+    val ktl_JOIN_om_WHERE_THIS_USER = s"""keep_to_library ktl inner join organization_membership om on (ktl.organization_id = om.organization_id) where om.state = 'active' and ktl.state = 'active' and (ktl.visibility = 'published' or ktl.visibility = 'organization') and om.user_id = $userId"""
+    val ktu_WHERE_THIS_USER = s"""keep_to_user ktu where ktu.state = 'active' and ktu.user_id = $userId"""
+
+    def getLastActivityAt(keepId: Id[Keep]): Option[DateTime] = sql"""SELECT last_activity_at from bookmark where id = $keepId """.as[Option[DateTime]].first
+
+    def getKeepIdAndLastActivityAt(id: ExternalId[Keep]): Option[(Id[Keep], DateTime)] = {
+      for {
+        keep <- getOpt(id)
+        keepId <- keep.id
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala]: this seems a bit odd, but I'm not necessarily opposed to it

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala]: just a little misleading. how about `keepId = keep.id.get`?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 2]
```diff
@@ -675,6 +676,64 @@ class KeepRepoImpl @Inject() (
...
+      case None => "true"
+      case Some((keepId, before)) => s"last_activity_at <= '$before' AND keep_id < $keepId"
+    }
+
+    val last_activity_at_AFTER = afterIdOpt.flatMap(getKeepIdAndLastActivityAt) match {
+      case None => "true"
+      case Some((_, after)) => s"last_activity_at > '$after'"
+      // This is not strictly correct. It's not possible to call after a keep, and get other keeps kept in the same ms.
+      // Fortunately, ending in this state where you have a keep id and need ones that happened after (and they happened in the same ms)
+      // is nearly impossible. We can't use IDs as tie breakers because old IDs may get updated added_at fields.
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala]: You can get rid of this comment if you add a `OR (last_activity_at = $after AND id > $keepId)`

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 3]
```diff
@@ -675,6 +676,64 @@ class KeepRepoImpl @Inject() (
...
+    }
+
+    val keepsAndLastActivityAt = sql"""
+      SELECT k.id as keep_id, max(k.most_recent_activity) as last_activity_at
+      FROM (#$fromQuery) as k
+      GROUP BY k.id
+      HAVING #$last_activity_at_AFTER AND #$last_activity_at_BEFORE
+      ORDER BY last_activity_at DESC, keep_id DESC
+      LIMIT $limit
+    """.as[(Id[Keep], DateTime)].list
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:725]: Any idea how slow this query is? Have you created the necessary indices?

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala]: :+1:

**Comment by ryanpbrewster** [comment]
lgtm. I always forget how involved the feed query is. Hopefully we can deprecate the `addedAt` feed and replace it with this one soon.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 3 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:725]: adding them now :) thank you

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 2 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala]: :+1:

**Comment by camhashemi** [comment]
yeah, and it may get worse. we want to start filtering by whether the keep is an image, video, or a given domain. Since these are likely few and far between, we have to do the filtering before sorting and taking limit, meaning `bookmark` may have to get involved in the query. Otherwise, we could just bump up the `limit` and hope we get enough images to populate a feed, still thinking about it.

**Pull request merged**